benchmark/single-source/StringReplaceSubrange.swift

```diff
@@ -0,0 +1,116 @@
+//===--- StringReplaceSubrange.swift -------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import TestsUtils
+
+let tags: [BenchmarkCategory] = [.validation, .api, .String]
+
+public let StringReplaceSubrange = [
+  BenchmarkInfo(
+    name: "Str.replaceSubrange.SmallLiteral.String",
+    runFunction: { replaceSubrange($0, "coffee", with: "t") },
+    tags: tags
+  ),
+  BenchmarkInfo(
+    name: "Str.replaceSubrange.LargeManaged.String",
+    runFunction: { replaceSubrange($0, largeManagedString, with: "t") },
+    tags: tags,
+    setUpFunction: setupLargeManagedString
+  ),
+  BenchmarkInfo(
+    name: "Str.replaceSubrange.SmallLiteral.Substr",
+    runFunction: { replaceSubrange($0, "coffee", with: getSubstring("t")) },
+    tags: tags
+  ),
+  BenchmarkInfo(
+    name: "Str.replaceSubrange.LargeManaged.Substr",
+    runFunction: { replaceSubrange($0, largeManagedString, with: getSubstring("t")) },
+    tags: tags,
+    setUpFunction: setupLargeManagedString
+  ),
+  BenchmarkInfo(
+    name: "Str.replaceSubrange.SmallLiteral.ArrChar",
+    runFunction: { replaceSubrange($0, "coffee", with: getArrayCharacter(Array<Character>(["t"]))) },
+    tags: tags
+  ),
+  BenchmarkInfo(
+    name: "Str.replaceSubrange.LargeManaged.ArrChar",
+    runFunction: { replaceSubrange($0, largeManagedString, with: getArrayCharacter(Array<Character>(["t"]))) },
+    tags: tags,
+    setUpFunction: setupLargeManagedString
+  ),
+  BenchmarkInfo(
+    name: "Str.replaceSubrange.SmallLiteral.RepeatedChar",
+    runFunction: { replaceSubrange($0, "coffee", with: getRepeatedCharacter(repeatedCharacter)) },
+    tags: tags
+  ),
+  BenchmarkInfo(
+    name: "Str.replaceSubrange.LargeManaged.RepeatedChar",
+    runFunction: { replaceSubrange($0, largeManagedString, with: getRepeatedCharacter(repeatedCharacter)) },
+    tags: tags,
+    setUpFunction: setupLargeManagedString
+  ),
+]
+
+// MARK: - Privates for String
+
+private var largeManagedString: String = {
+    return getString("coffee\u{301}coffeecoffeecoffeecoffee")
+}()
```

Comment:
`\u{301}` was suggested to be added to the test string when I paired with Michael. @milseman would you mind reminding me the reason for this? I don't remember it now 😅

Reply:
I'll guess it was in order for the string to be in particular [normalization form](https://unicode.org/reports/tr15/). @milseman Do you want to vary the benchmarks also for different normalization forms? [SR-8905](https://bugs.swift.org/browse/SR-8905) doesn't mention that…

Reply:
I was wondering if the result of testing with the string "coffeécoffeecoffeecoffeecoffee" would be different from the one with "coffeecoffeecoffeecoffeecoffee" 🤔 If there is distinct difference, maybe we could add two benchmarks for the one with the acute accent character and the other one without it?

Reply:
Grapheme segmentation is relevant to this benchmark, but not normalization (since there's no comparison). The difference is that the precomposed representation is a single scalar per grapheme cluster, while the decomposed (multi-scalar) form is not. The single-scalar one will hit our grapheme breaking fast-paths while the multi-scalar one will call out to ICU. Alternatively, you could use other kinds of multi-scalar graphemes clusters, such as complex emoji. I just mentioned `"\u{301}"` because you can just stick it after an `"e"` to get the same effect.